# Insolvencias LibreBOR


## Monthly Bankruptcies

2021 does not have a peak in bankruptcies, actually they seem to be ciclical with a decreasing pattern.

In [35]:
import pandas as pd
import ast
import altair as alt
from vega_datasets import data

filename = "data/insolvencias_2021.04.20"
df = pd.read_csv(filename + '.csv', low_memory = False)

df["day"]= [x[:7] for x in df["date"]]
df.sort_values(by = ["day"],inplace = True)
df.head(10)

df_dates = pd.DataFrame(df.groupby(['day']).size()).reset_index("day")

df_dates_cn = df_dates.rename(columns = {0:"recuento_insolvencias", "day":"month"})



In [36]:
###   GENERATE INTERACTIVE PLOT   ###
source = df_dates_cn

# Create a selection that chooses the nearest point & selects based on x-value
nearest = alt.selection(type='single', nearest=True, on='mouseover',
                        fields=['month'], empty='none')

# The basic line
line = alt.Chart(source).mark_line(interpolate='basis').encode(
    x='month:T',
    y='recuento_insolvencias:Q'
)

# Transparent selectors across the chart. This is what tells us
# the x-value of the cursor
selectors = alt.Chart(source).mark_point().encode(
    x='month:T',
    opacity=alt.value(0),
).add_selection(
    nearest
)

# Draw points on the line, and highlight based on selection
points = line.mark_point().encode(
    opacity=alt.condition(nearest, alt.value(1), alt.value(0))
)

# Draw text labels near the points, and highlight based on selection
text = line.mark_text(align='left', dx=2, dy=-2).encode(
    text=alt.condition(nearest, 'recuento_insolvencias:Q', alt.value(' '))
)

# Draw a rule at the location of the selection
rules = alt.Chart(source).mark_rule(color='gray').encode(
    x='month:T',
).transform_filter(
    nearest
)

# Put the five layers into a chart and bind the data
chart2 = alt.layer(
    line, selectors, points, rules, text
).properties(
    width=600, height=300
)

#chart2.save(export_file + '_interactive.html')
chart2
#print("Interactive plot has been successfully saved at " + export_file + '_interactive.html')



alt.LayerChart(...)

## Monthly Bankruptcies

Note that August traditionally is a constantly year minimum:

In [30]:
###   GENERATE INTERACTIVE PLOT   ###
source = df_dates_cn

# Create a selection that chooses the nearest point & selects based on x-value
nearest = alt.selection(type='single', nearest=True, on='mouseover',
                        fields=['month'], empty='none')

# The basic line
line = alt.Chart(source).mark_line(interpolate='basis').encode(
    x='month:T',
    y='recuento_insolvencias:Q'
)

# Transparent selectors across the chart. This is what tells us
# the x-value of the cursor
selectors = alt.Chart(source).mark_point().encode(
    x='month:T',
    opacity=alt.value(0),
).add_selection(
    nearest
)

# Draw points on the line, and highlight based on selection
points = line.mark_point().encode(
    opacity=alt.condition(nearest, alt.value(1), alt.value(0))
)

# Draw text labels near the points, and highlight based on selection
text = line.mark_text(align='left', dx=2, dy=-2).encode(
    text=alt.condition(nearest, 'month:T', alt.value(' '))
)

# Draw a rule at the location of the selection
rules = alt.Chart(source).mark_rule(color='gray').encode(
    x='month:T',
).transform_filter(
    nearest
)

# Put the five layers into a chart and bind the data
chart2 = alt.layer(
    line, selectors, points, rules, text
).properties(
    width=600, height=300
)

#chart2.save(export_file + '_interactive.html')
chart2
#print("Interactive plot has been successfully saved at " + export_file + '_interactive.html')



alt.LayerChart(...)

In [37]:
df_provinces = pd.DataFrame(df.groupby(['province']).size().reset_index("province"))
df_provinces = df_provinces.rename(columns = {0:"count_of_bankruptcies"})
df_provinces = df_provinces.sort_values("count_of_bankruptcies", ascending = False)
df_provinces

province  count_of_bankruptcies
7                Barcelona                  12176
47                Valencia                   5524
33                  Murcia                   5373
51                Zaragoza                   4757
34                  Málaga                   4552
5                 Asturias                   3154
11               Castellón                   1957
46                  Toledo                   1759
38              Pontevedra                   1576
2                 Alicante                   1509
30                  Lleida                   1229
1                 Albacete                   1208
40  Santa Cruz De Tenerife                   1189
27                La Rioja                   1101
20                 Granada                    738
18                Gipuzkoa                    689
42                 Sevilla                    680
32                  Madrid                    637
19                  Girona                    551
28              Las Palmas                    514
0                 A Coruña                    445
22               Guipúzcoa                    372
14                  Cuenca                    291
15                 Cáceres                    233
44               Tarragona                    193
53                   Ávila                    189
41                 Segovia                    151
6                  Badajoz                    144
50                  Zamora                    129
25           Illes Balears                    127
37                Palencia                     90
16                   Cádiz                     90
29                    León                     85
3                  Almería                     82
13             Ciudad Real                     79
24                  Huesca                     79
26                    Jaén                     76
43                   Soria                     74
48              Valladolid                     72
36                 Ourense                     62
8                  Bizkaia                     61
10               Cantabria                     58
49                 Vizcaya                     49
39               Salamanca                     38
45                  Teruel                     27
9                   Burgos                     26
4              Araba/Álava                     21
35                 Navarra                     17
52                   Álava                     14
23                  Huelva                     12
17                 Córdoba                     11
31                    Lugo                      9
21             Guadalajara                      5
12                   Ceuta                      2

## Bankruptcies by Province

It seems that the data is very focused on Barcleona:


In [33]:
import altair as alt
import pandas as pd


alt.Chart(df_provinces).mark_bar().encode(
    y='count_of_bankruptcies:Q',
    x=alt.Y('province:N', sort='-y'),
    tooltip = ['province','count_of_bankruptcies']
)

alt.Chart(...)

## Bankruptcies by Province

- Barcelona keeps most of the Bankruptcies, it doubles the following Province (Valencia).

- Barcelona has almost 20 times more bankruptcies than madrid (*Madrid has a total of 637 bankruptcies vs 12.176 from Barcleona*)

- Many Provinces have irrelevant amount of bankruptcies.


## Top Province Bankruptcy Evolution

Every Province has less bankruptcies in the later years (2019-2021) and concentrate maximums of bankruptcies on 2013-2019.

It looks like in early years (2013-2019) only data from Barcleona was collected, or there is some bias on the data extraction, or lack of data to complete the later years.


In [38]:
#Get top 10 provinces

top_5_prov = df_provinces["province"][:5]


df_provinces_month = pd.DataFrame(df.groupby(['province', 'day']).size().reset_index())
df_provinces_month = df_provinces_month.rename(columns = {0:"count_of_bankruptcies"})
df_provinces_month.sort_values(["day","count_of_bankruptcies"], ascending = True)

df_provinces_month_rn = df_provinces_month.rename(columns = {"province":"category", "day":"x", "count_of_bankruptcies":"y"})
df_provinces_month_rn

df_provinces_month_top5 = df_provinces_month_rn[df_provinces_month_rn["category"].isin(list(top_5_prov))]
df_provinces_month_top5

category        x   y
466   Barcelona  2009-06   2
467   Barcelona  2010-07   1
468   Barcelona  2011-04   1
469   Barcelona  2011-08   1
470   Barcelona  2011-11   1
...         ...      ...  ..
3298   Zaragoza  2020-12   9
3299   Zaragoza  2021-01   8
3300   Zaragoza  2021-02  29
3301   Zaragoza  2021-03  19
3302   Zaragoza  2021-04   3

[551 rows x 3 columns]

In [39]:
import altair as alt
import pandas as pd
import numpy as np

np.random.seed(42)
source = pd.DataFrame(np.cumsum(np.random.randn(100, 3), 0).round(2),
                    columns=['A', 'B', 'C'], index=pd.RangeIndex(100, name='x'))
source = source.reset_index().melt('x', var_name='category', value_name='y')

source = df_provinces_month_top5


# Create a selection that chooses the nearest point & selects based on x-value
nearest = alt.selection(type='single', nearest=True, on='mouseover',
                        fields=['x'], empty='none')

# The basic line
line = alt.Chart(source).mark_line(interpolate='basis').encode(
    x='x:T',
    y='y:Q',
    color='category:N'
)

# Transparent selectors across the chart. This is what tells us
# the x-value of the cursor
selectors = alt.Chart(source).mark_point().encode(
    x='x:T',
    opacity=alt.value(0),
).add_selection(
    nearest
)

# Draw points on the line, and highlight based on selection
points = line.mark_point().encode(
    opacity=alt.condition(nearest, alt.value(1), alt.value(0))
)

# Draw text labels near the points, and highlight based on selection
text = line.mark_text(align='left', dx=5, dy=-5).encode(
    text=alt.condition(nearest, 'y:Q', alt.value(' '))
)

# Draw a rule at the location of the selection
rules = alt.Chart(source).mark_rule(color='gray').encode(
    x='x:T',
).transform_filter(
    nearest
)

# Put the five layers into a chart and bind the data
alt.layer(
    line, selectors, points, rules, text
).properties(
    width=600, height=300
)

alt.LayerChart(...)

## Top Province Bankruptcy Evolution

Every province has a decreasing trend, Barcelona is suspiciously high.

In [40]:
import altair as alt
import pandas as pd
import numpy as np

np.random.seed(42)
source = pd.DataFrame(np.cumsum(np.random.randn(100, 3), 0).round(2),
                    columns=['A', 'B', 'C'], index=pd.RangeIndex(100, name='x'))
source = source.reset_index().melt('x', var_name='category', value_name='y')

source = df_provinces_month_rn


# Create a selection that chooses the nearest point & selects based on x-value
nearest = alt.selection(type='single', nearest=True, on='mouseover',
                        fields=['x'], empty='none')

# The basic line
line = alt.Chart(source).mark_line(interpolate='basis').encode(
    x='x:T',
    y='y:Q',
    color='category:N'
)

# Transparent selectors across the chart. This is what tells us
# the x-value of the cursor
selectors = alt.Chart(source).mark_point().encode(
    x='x:T',
    opacity=alt.value(0),
).add_selection(
    nearest
)

# Draw points on the line, and highlight based on selection
points = line.mark_point().encode(
    opacity=alt.condition(nearest, alt.value(1), alt.value(0))
)

# Draw text labels near the points, and highlight based on selection
text = line.mark_text(align='left', dx=5, dy=-5).encode(
    text=alt.condition(nearest, 'y:Q', alt.value(' '))
)

# Draw a rule at the location of the selection
rules = alt.Chart(source).mark_rule(color='gray').encode(
    x='x:T',
).transform_filter(
    nearest
)

# Put the five layers into a chart and bind the data
alt.layer(
    line, selectors, points, rules, text
).properties(
    width=600, height=300
)

alt.LayerChart(...)